### PR TITLE
feat: virtual wallet support + Metamask integration

### PR DESCRIPTION
**Pull request by naorye**
Virtual wallet is a wallet instance that has no implementation, which will be loaded upon calling to `enable` method.
Virtual wallet include two important methods:
1. `hasSupport()` - async, return true whether this wallet is supported
2. `loadWallet()` - async, download and return the actual wallet instance
Wallet vendors that requires additional code will be supported in get-starknet via virtual wallet.

For example, in Metamask integration, `hasSupport()` returns true if MetaMask extension exists and if there is a snap support. In such case, MetaMask will be an option for the available wallets. Once the user will choose Metamask and will enable it, the actual wallet implementation will be downloaded and return to the user.

### CODE REVIEW

**Comment by stanleyyconsensys** [comment]
The URL : https://s3.eu-central-1.amazonaws.com/dev.snaps.consensys.io/get-starknet/remoteEntry.js is developed base on current master/main branch of get-starknet 

hence it will not work to the get-starknet:develop branch

**Comment by naorye2** [comment]
@stanleyyconsensys 
Let me know which url to use.

File: packages/core/src/wallet/virtualWallets/index.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import type { VirtualWallet } from "../../types"
+import { metaMaskVirtualWallet } from "./metaMaskVirtualWallet"
+import type { StarknetWindowObject } from "starknet-types"
```
Comment (dan-ziv) [inline packages/core/src/wallet/virtualWallets/index.ts:3]:
```suggestion
import type { StarknetWindowObject } from "@starknet-io/types-js"
```

Comment (naorye2) [reply packages/core/src/wallet/virtualWallets/index.ts]: Thanks, fixed.

**Review by dan-ziv — APPROVED** [review]

File: packages/core/src/wallet/virtualWallets/metaMaskVirtualWallet.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        resolve(provider)
+        handled = true
+      }
+    }
+
+    if (typeof windowObject.addEventListener === "function") {
+      windowObject.addEventListener(
+        "eip6963:announceProvider",
+        handleEIP6963Provider,
+      )
```
Comment (avimak) [inline packages/core/src/wallet/virtualWallets/metaMaskVirtualWallet.ts:38]: this one could be registered once for all providers, right?

File: packages/core/src/wallet/virtualWallets/metaMaskVirtualWallet.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      if (!handled) {
+        resolve(null)
+      }
+    }, timeout)
+
+    // Notify event listeners and other parts of the dapp that a provider is requested.
+    if (typeof windowObject.dispatchEvent === "function") {
+      windowObject.dispatchEvent(new Event("eip6963:requestProvider"))
+    }
+  })
```
[the remark this answers is not in the document]
Comment (khanti42) [reply packages/core/src/wallet/virtualWallets/metaMaskVirtualWallet.ts:51]: This standard is designed for ethereum wallet detection so I am not sure it's interesting for other providers on Starknet. This basically depends on how multi wallet discovery is handled currently on starknet but adopting the EIP6963 standard could maybe be an option ?

**Pull request merged**